app/src/main/java/org/kiwix/kiwixmobile/wifi_hotspot/HotspotService.java

```diff
@@ -0,0 +1,224 @@
+  private static final String TAG = "HotspotService";
+  private BroadcastReceiver stopReceiver;
+  private NotificationManager notificationManager;
+  private NotificationCompat.Builder builder;
+  private ZimHostCallbacks zimHostCallbacks;
+  private final IBinder serviceBinder = new HotspotBinder();
+
+  @Inject
+  WebServerHelper webServerHelper;
+
+  @Inject
+  WifiHotspotManager hotspotManager;
+
+  @Override public void onCreate() {
+    KiwixApplication.getApplicationComponent()
+        .serviceComponent()
+        .service(this)
+        .build()
+        .inject(this);
+    super.onCreate();
+
+    hotspotManager.registerListener(this);
+
+    stopReceiver = new BroadcastReceiver() {
+      @Override
+      public void onReceive(Context context, Intent intent) {
+        if (intent != null && intent.getAction().equals(ACTION_STOP)) {
+          stopHotspotAndDismissNotification();
+        }
+      }
+    };
+
+    registerReceiver(stopReceiver, new IntentFilter(ACTION_STOP));
+
+    notificationManager = (NotificationManager) getSystemService(Context.NOTIFICATION_SERVICE);
+  }
+
+  @Override public int onStartCommand(@NonNull Intent intent, int flags, int startId) {
+    switch (intent.getAction()) {
+
+      case ACTION_TOGGLE_HOTSPOT:
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
+          if (hotspotManager.isHotspotStarted()) {
+            stopHotspotAndDismissNotification();
+          } else {
+            zimHostCallbacks.requestLocationAccess();
+          }
+        }
+        break;
+
+      case ACTION_LOCATION_ACCESS_GRANTED:
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
+          hotspotManager.turnOnHotspot();
+        }
+        break;
+
+      case ACTION_START_SERVER:
+        if (webServerHelper.startServerHelper(
+            intent.getStringArrayListExtra(SELECTED_ZIM_PATHS_KEY))) {
+          zimHostCallbacks.onServerStarted(ServerUtils.getSocketAddress());
+          if (Build.VERSION.SDK_INT < Build.VERSION_CODES.O) {
+            startForegroundNotificationHelper();
+          }
+          Toast.makeText(this, R.string.server_started__successfully_toast_message,
+              Toast.LENGTH_SHORT).show();
+        } else {
+          zimHostCallbacks.onServerFailedToStart();
+          if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
+            stopForeground(true);
+            stopSelf();
+            notificationManager.cancel(HOTSPOT_NOTIFICATION_ID);
+          }
+        }
+
+        break;
+
+      case ACTION_STOP_SERVER:
+        stopHotspotAndDismissNotification();
+        break;
+      default:
+        break;
+    }
+    return START_NOT_STICKY;
+  }
+
+  @Nullable @Override public IBinder onBind(@Nullable Intent intent) {
+    return serviceBinder;
+  }
+
+  private Notification buildForegroundNotification(String status) {
+    Log.v(TAG, "Building notification " + status);
+    builder = new NotificationCompat.Builder(this);
```

Comment:
just use the builder, also why is the builder a field?
Code I want to see
```
// create your intents up top
return new NotificationCompat.Builder(this)
                    .setContentTitle(R.string.hotspot_notification_content_title)
                      //set the rest of the stuff
                      .build();
```

Also as I said on a much older comment I am still awaiting all notification stuff to be extracted to a separate class

Reply:
Fixed this and also added all notification related stuff into a separate class.
d1572c142d6b301ade94e6cdcf38f947dad4e925